### PR TITLE
1.3.0

### PR DESCRIPTION
**Pull request by gusaaaaa**
## References

- [Jira Release](https://focusuy.atlassian.net/projects/BMC2/versions/10275/tab/release-report-all-issues)

## Changes

**New Functionality**

- [BMC2-898](https://focusuy.atlassian.net/browse/BMC2-898) Integrate I3C Target Reset Support into SupernovaController
- [BMC2-953](https://focusuy.atlassian.net/browse/BMC2-953) Implement hot-join functionality in SupernovaController
- [BMC2-1057](https://focusuy.atlassian.net/browse/BMC2-1057) Add methods to change Static Address, BCR, DCR and PID for the I3C Target in the SupernovaController

**Interaction**

- [BMC2-1257](https://focusuy.atlassian.net/browse/BMC2-1257) Enhance IBI example in SupernovaController

**API Enhancements**

- [BMC2-825](https://focusuy.atlassian.net/browse/BMC2-825) Add find\_target\_device\_by\_pid utility method to the SupernovaI3CBlockingInterface class
- [BMC2-840](https://focusuy.atlassian.net/browse/BMC2-840) Add wrapper of getConnectedSupernovaDevicesList from the SupernovaSDK to the SupernovaController

**Stability and Reliability**

- [BMC2-1256](https://focusuy.atlassian.net/browse/BMC2-1256) Enhance Error Debugging for I2C Failures in SupernovaController
- [BMC2-1293](https://focusuy.atlassian.net/browse/BMC2-1293) Update SPI WhoAmI transfer test to support Simulator
- [BMC2-1376](https://focusuy.atlassian.net/browse/BMC2-1376) Remove target reset tests that will be included later in SupernovaController v1.4.0
- [BMC2-861](https://focusuy.atlassian.net/browse/BMC2-861) Integrate simulators package
- [BMC2-1035](https://focusuy.atlassian.net/browse/BMC2-1035) Update BinhoSimulator Dependency in SupernovaController setup.py
- [BMC2-1378](https://focusuy.atlassian.net/browse/BMC2-1378) Add debugging configuration for vscode

**Compatibility with Supernova SDK 2.2.0**

- [BMC2-1261](https://focusuy.atlassian.net/browse/BMC2-1261) Update SDK version in SupernovaController dependencies
- [BMC2-1014](https://focusuy.atlassian.net/browse/BMC2-1014) Update BinhoSimulators test.py to correctly test the SDK 2.2.0 behavior and simulators 0.2.1

**Learnability**

- [BMC2-1262](https://focusuy.atlassian.net/browse/BMC2-1262) Enrich Documentation for SupernovaI2CBlockingInterface Class
- [BMC2-1374](https://focusuy.atlassian.net/browse/BMC2-1374) Update data_files array in setup.py according to new additions and nomenclature

### Run tests

⚠️ Use a Supernova with FW v2.3.0 or v2.4.0 ⚠️ 
After activating your virtual environment with python 3.9 or greater, run `pip install .`

Run tests with: `python tests\test.py`

**Expected results for:**
<details><summary>Real device</summary>
<p>

```
ss..........ssssssss.s..s.....s..
----------------------------------------------------------------------
Ran 33 tests in 4.712s

OK (skipped=13)
```

</p>
</details> 

<details><summary>Simulator</summary>
<p>

```
...........s..............EE.EEs.ss
======================================================================
ERROR: test_spi_controller_get_parameters (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 190, in init_bus
    responses = self.controller.sync_submit([
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 139, in sync_submit
    raise error_info
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 45, in sequence_runner
    func(request_state['transfer_ids'][current_index])
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 191, in <lambda>
    lambda transfer_id: self.driver.spiControllerInit(id=transfer_id, bitOrder=self.bit_order, mode=self.mode, dataWidth=self.data_width, chipSelect=self.chip_select, chipSelectPol=self.chip_select_pol, frequency=self.frequency)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\binhosimulators\helpers\wrappers\supernova_wrappers.py", line 38, in wrapper
    response = self.get_spi_transfer_response_template(args[0], func.__name__.replace('_', ' ').upper(), 0, 0)
IndexError: tuple index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 490, in test_spi_controller_get_parameters
    spi_controller.init_bus()
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 194, in init_bus
    raise BackendError(original_exception=e) from e
supernovacontroller.errors.exceptions.BackendError: An error occurred in the backend: tuple index out of range

======================================================================
ERROR: test_spi_controller_init_bus (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 190, in init_bus
    responses = self.controller.sync_submit([
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 139, in sync_submit
    raise error_info
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 45, in sequence_runner
    func(request_state['transfer_ids'][current_index])
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 191, in <lambda>
    lambda transfer_id: self.driver.spiControllerInit(id=transfer_id, bitOrder=self.bit_order, mode=self.mode, dataWidth=self.data_width, chipSelect=self.chip_select, chipSelectPol=self.chip_select_pol, frequency=self.frequency)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\binhosimulators\helpers\wrappers\supernova_wrappers.py", line 38, in wrapper
    response = self.get_spi_transfer_response_template(args[0], func.__name__.replace('_', ' ').upper(), 0, 0)
IndexError: tuple index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 465, in test_spi_controller_init_bus
    (success, _) = spi_controller.init_bus()
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 194, in init_bus
    raise BackendError(original_exception=e) from e
supernovacontroller.errors.exceptions.BackendError: An error occurred in the backend: tuple index out of range

======================================================================
ERROR: test_spi_controller_set_parameters (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 190, in init_bus
    responses = self.controller.sync_submit([
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 139, in sync_submit
    raise error_info
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 45, in sequence_runner
    func(request_state['transfer_ids'][current_index])
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 191, in <lambda>
    lambda transfer_id: self.driver.spiControllerInit(id=transfer_id, bitOrder=self.bit_order, mode=self.mode, dataWidth=self.data_width, chipSelect=self.chip_select, chipSelectPol=self.chip_select_pol, frequency=self.frequency)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\binhosimulators\helpers\wrappers\supernova_wrappers.py", line 38, in wrapper
    response = self.get_spi_transfer_response_template(args[0], func.__name__.replace('_', ' ').upper(), 0, 0)
IndexError: tuple index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 476, in test_spi_controller_set_parameters
    spi_controller.init_bus()
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 194, in init_bus
    raise BackendError(original_exception=e) from e
supernovacontroller.errors.exceptions.BackendError: An error occurred in the backend: tuple index out of range

======================================================================
ERROR: test_spi_sim_transfer (__main__.TestSupernovaController)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 190, in init_bus
    responses = self.controller.sync_submit([
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 139, in sync_submit
    raise error_info
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\transfer_controller\__init__.py", line 45, in sequence_runner
    func(request_state['transfer_ids'][current_index])
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 191, in <lambda>
    lambda transfer_id: self.driver.spiControllerInit(id=transfer_id, bitOrder=self.bit_order, mode=self.mode, dataWidth=self.data_width, chipSelect=self.chip_select, chipSelectPol=self.chip_select_pol, frequency=self.frequency)
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\binhosimulators\helpers\wrappers\supernova_wrappers.py", line 38, in wrapper
    response = self.get_spi_transfer_response_template(args[0], func.__name__.replace('_', ' ').upper(), 0, 0)
IndexError: tuple index out of range

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\jpons\Binho\BMC\SupernovaController\tests\test.py", line 540, in test_spi_sim_transfer
    spi_controller.init_bus()
  File "C:\Users\jpons\Binho\BMC\SupernovaController\env\lib\site-packages\supernovacontroller\sequential\spi_controller.py", line 194, in init_bus
    raise BackendError(original_exception=e) from e
supernovacontroller.errors.exceptions.BackendError: An error occurred in the backend: tuple index out of range

----------------------------------------------------------------------
Ran 35 tests in 0.114s

FAILED (errors=4, skipped=4)
```
**Note:** These errors will disappear when the bug reported [here](https://focusuy.atlassian.net/browse/BMC2-1331) is addressed.

</p>
</details> 

### Run examples

Run the most relevant examples according what changes are being introduced in this PR:
1. `python examples/hot_join_example.py`
2. `python examples/i3c_ibi_example.py`
3. `python examples/basic_spi_controller_example.py`

**Expected results:**

<details><summary>1. `hot_join_example.py`</summary>
<p>

```
Initializing and setting up the I3C peripheral
No targets joined the I3C bus via the ENTDAA CCC
```
And after powering the PIC18F16Q20, a notification should be printed: together with the updated table:
```
NOTIFICATION: New device added via hot-join procedure: {'dynamic_address': 8, 'bcr': 30, 'dcr': 198, 'pid': ['0x06', '0x9a', '0x00', '0x00', '0x00', '0x00']}
The target table is: [{'static_address': 0, 'dynamic_address': 8, 'bcr': 30, 'dcr': 198, 'pid': ['0x00', '0x00', '0x00', '0x00', '0x9a', '0x06']}]

```
</p>
</details> 

<details><summary>2. `i3c_ibi_example.py`</summary>
<p>

```
Connected device info: {'static_address': 0, 'dynamic_address': 8, 'bcr': 39, 'dcr': 160, 'pid': ['0x04', '0x6a', '0x00', '0x00', '0x00', '0x00']}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
NOTIFICATION: New IBI request -> {'dynamic_address': 8, 'controller_response': 'IBI_ACKED_WITH_PAYLOAD', 'mdb': 2, 'payload': [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
```

</p>
</details> 

<details><summary>3. `basic_spi_controller_example.py`</summary>
<p>

```
Opening Supernova host adapter device and getting access to the SPI protocol interface...
Initializing SPI peripheral.
The configured parameters are:
Bit order: LSB, Mode: MODE_0, Data width: _8_BITS_DATA, Chip select: CHIP_SELECT_0, Chip select polarity: ACTIVE_LOW, Frequency: 10000000
Changing the bit order to MSB first.
The configured parameters are:
Bit order: MSB, Mode: MODE_0, Data width: _8_BITS_DATA, Chip select: CHIP_SELECT_0, Chip select polarity: ACTIVE_LOW, Frequency: 10000000
Reading manufacturer ID
Received data: ['0x00', '0x04', '0x7f', '0x03', '0x02']
Enabling write operation
Write operation enable
Writing [0x2A 0x2B] to 0x0000 memory address
Write operation completed successfully
Reading 0x0000 memory address
Received data: ['0x00', '0x00', '0x00', '0x2a', '0x3b']
Data written and read match correctly
```

</p>
</details>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -524,6 +563,7 @@ After installing the `SupernovaController` package, you can further explore its
 - **Basic I2C Example (`basic_i2c_example.py`):** Get started with fundamental I2C operations.
 - **Basic UART Example (`basic_uart_example.py`):** Try out the UART protocol Hands-On.
 - **Basic SPI Controller Example (`basic_spi_controller_example.py`):** Explore the fundamental SPI controller operations communicating with a SPI Target device.
+- **Hot-join example(`hot_join_example.py`):** Understand how to handle the hot-join procedure in I3C.
 - **IBI Example (`ibi_example.py`):** Understand handling In-Band Interrupts (IBI) in I3C.
```
Comment (Salonso1602) [inline README.md:567]: This file changed names

**Review by Salonso1602 — APPROVED** [review]
Left a minor comments
look good

**Review by gusaaaaa — COMMENTED** [review]
Looks good to me!

File: examples/i3c_target_set_ids.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    - Assign dynamic addresses using RSTDAA and SETDASA.
+    - Close Device Connection: Close the connection to the Supernova devices.
+    """
+
+    print("-----------------------")
+    print("Initialize Supernovas")
+    print("-----------------------")
+
+    target_device = SupernovaDevice()
+    info = target_device.open(usb_address ="\\\\?\\HID#VID_1FC9&PID_82FC#6&bb45e52&1&0000#{4d1e55b2-f16f-11cf-88cb-001111000030}")
```
Comment (felipemoran93) [inline examples/i3c_target_set_ids.py:33]: are we cool with hard-coded addresses? If not, let's make tickets to address this in the future.

**Review by felipemoran93 — APPROVED** [review]
LGTM! I just left a little question

File: examples/i3c_target_set_ids.py (hunk 2 above)
Comment (pons93) [reply examples/i3c_target_set_ids.py:33]: Given that currently we have the `openAllConnectedSupernovaDevices()` method in `SupernovaDevice` class, we should use it across all examples. I just created [this](https://focusuy.atlassian.net/browse/BMC2-1402) ticket to address that. Thanks for the observation!

File: examples/i3c_target_set_ids.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    - Discover Targets on I3C Bus: Fetch a list of devices present on the I3C bus.
+    - Assign dynamic addresses using RSTDAA and SETDASA.
+    - Close Device Connection: Close the connection to the Supernova devices.
+    """
+
+    print("-----------------------")
+    print("Initialize Supernovas")
+    print("-----------------------")
+
+    target_device = SupernovaDevice()
```
[the remark this answers is not in the document]
Comment (pons93) [reply examples/i3c_target_set_ids.py:32]: As I commented [here](https://github.com/binhollc/SupernovaController/pull/40#discussion_r1647804552), I've created [this](https://focusuy.atlassian.net/browse/BMC2-1402) ticket to address this situation. Thanks Fausto!

File: README.md [hunk 4]
```diff
@@ -217,8 +217,47 @@ In an I3C bus, the Supernova can act either as a controller or as a target.
...
    The memory layout field can take `MEM_1_BYTE`, `MEM_2_BYTES` or `MEM_4_BYTES` value.
-        
-2. ***Set Supernova configuration:***
+
+2. ***Set PID:***
+
+    Sets the PID of the Supernova acting as an I3C target via USB:
+
+    ```python
+   success, error = device.set_pid([0x02, 0x03, 0x04, 0x05, 0x06, 0x07])
```
[the remark this answers is not in the document]
Comment (pons93) [reply README.md:226]: Good observation! I reported it [here](https://focusuy.atlassian.net/browse/BMC2-1403) to include it for SC v1.4.0.

**Pull request closed without merging**